README.md
![alt text][logo]

[logo]: https://github.com/MagnumOpuses/project-meta/blob/master/img/jobtechdev_black.png "JobTech dev logo"

[A JobTech Project](https://www.jobtechdev.se)

# AF Connect Module

- [Introduction](#introduction)
  - [Live use-case demo](#live-use-case-demo)
  - [AF-Connect communication diagram](#af-connect-communication-diagram)
  - [AF-Connect Module sequence diagram](#af-connect-module-sequence-diagram)
- [Getting started](#getting-started)
  - [Using docker](#using-docker)
  - [Basic front-end example](#basic-front-end-example)
- [AF-Connect integration environment](#af-connect-integration-environment)
- [Configuration](#configuration)
- [Build from source](#build-from-source)
- [Test](#test)
- [Error codes](#error-codes)
- [Built with](#built-with)
- [Versions, current dev state and future](#versions-current-dev-state-and-future)
- [Contributing](#contributing)
- [License](#license)
- [Acknowledgments](#acknowledgments)

## Introduction

The purpose of this module is to enable third-party services to easily integrate with Arbetsförmedlingen's AF-Connect infrastructure in order to obtain data/documents/certificates related to registered jobseekers and/or employers with their explicit consent.

This module provides an interactive button for you to install into your own service by following the guidelines below.

Once you've successfully installed this module, your website visitor(s) can initiate transaction of their personal data to your service by clicking the AF-Connect button.

When the user initiates the transaction of data, a new window will open up for the user to authenticate at Arbetsförmedlingen's website, select their profile/CV to share and finally provide their explicit consent allowing this data to be shared with the third-party service.

### Live use-case demo

A jobseeker fills out a job application form automatically by using AF-Connect.

![AF Connect user journey diagram v1](.github/screenshots/af-connect-user-journey-diagram-v1.png?raw=true)

Use the following test credentials listed below and experience the live demo here: [https://af-connect-demo.jobtechdev.se](https://af-connect-demo.jobtechdev.se)

```
Username: abc
Password: 123
```

### AF-Connect communication diagram

![AF Connect communication diagram v3](.github/screenshots/af-connect-communication-diagram-v3.png?raw=true)

### AF-Connect Module sequence diagram

![AF Connect Module sequence diagram v2](.github/screenshots/af-connect-module-sequence-diagram-v2.png?raw=true)

## Getting started

### Using Docker

We provide a docker release package for you to easily run the entire AF-Connect system locally in your development machine and effectively allowing you to both experience and evaluate whether this integration brings value to your product/service.

First and foremost, add the following host to your local `hosts` file:

```
127.0.0.1 af-connect.local
```

Download the docker release package from [af-connect-compose](https://github.com/MagnumOpuses/af-connect-compose), unpack the archive and start up all the pre-configured services with command: `docker-compose up`.

The following docker images will be installed:

- [jobtechdev/af-connect-demo:latest](https://hub.docker.com/r/jobtechdev/af-connect-demo)
- [jobtechdev/af-connect:latest](https://hub.docker.com/r/jobtechdev/af-connect)
- [jobtechdev/af-portability:latest](https://hub.docker.com/r/jobtechdev/af-portability)
- [jobtechdev/af-connect-outbox:latest](https://hub.docker.com/r/jobtechdev/af-connect-outbox)
- [jobtechdev/apimanager-mock:latest](https://hub.docker.com/r/jobtechdev/apimanager-mock)
- [jobtechdev/af-connect-mock:latest](https://hub.docker.com/r/jobtechdev/af-connect-mock)
- [redis:alpine](https://hub.docker.com/_/redis)

When all docker images have started up, visit your locally hosted AF-Connect Demo site at: [http://af-connect.local:3000](http://af-connect.local:3000)

### Basic front-end example

Now let's create a basic front-end example that utilizes the AF-Connect-Module.

Create a new directory in your filesystem and add a new file `index.html` with the following content:

```html
<html>
  <body>
    <script type="text/javascript">
      function onResponse(envelope, err) {
        if (err !== undefined) {
          alert("Error [" + err.code + "]: " + err.message);
          return;
        }

        // Envelope contains user CV and consent details.
        console.log(envelope);
      }
    </script>
    <div class="af-connect-module" data-on_response="onResponse"></div>
    <script src="af-connect-module.bundle.js"></script>
  </body>
</html>
```

The example consists of three parts.

- A globally defined callback for handling received data responses from AF-Connect.
- A DOM element which defines the configuration properties and location for the interactive button.
- The pre-compiled module script that will bring this all to life by generating the interactive button.

Download and unpack the latest version of the AF Connect Module from the [Releases section](https://github.com/MagnumOpuses/af-connect-module/releases). Copy the `af-connect-module.bundle.js` into the directory where your `index.html` resides.

Finally open up the `index.html` in your web browser and you'll find a blue "AF Connect" button.

![Example implementation result screenshot](https://github.com/MagnumOpuses/af-connect-module/blob/master/.github/screenshots/af-connect-module-button.png?raw=true)

By clicking the "AF Connect" button, a new tab/window will open up to show you an authentication page. This page is operating on mocked user data. To complete the authentication procedure, enter the following user credentials:

```
Username: abc
Password: 123
```

Upon successful authentication a handful of example profiles are presented to you. Please select the profile you wish to share with the third-party service and finally click the "Consent and close" button. The AF-Connect page closes and you're back to your `index.html` page again.

Open up the browser developer tools console and you will see that the `onResponse` callback have been called, and the retrieved envelope has been logged to the console.

:notebook: _Notice: Multiple AF Connect Modules can be added to a page through duplication of the DOM element._

## AF-Connect integration environment

In order to integrate your own website with AF-Connect integration environment you must apply/adjust the configuration properties as listed below.

| Configuration property      | Af-connect integration env.              | Description                        |
| --------------------------- | ---------------------------------------- | ---------------------------------- |
| data-af_connect_url         | https://af-connect-int.jobtechdev.se     | se [Configuration](#configuration) |
| data-af_portability_url     | https://af-portability-int.jobtechdev.se | se [Configuration](#configuration) |
| data-af_portability_api_key | dummykey                                 | se [Configuration](#configuration) |

Find all available configuration properties in section: [Configuration](#configuration)

## Configuration

The interactive button comes with pre-defined configuration defaults, but each property may be overridden as you see fit.

Here's an example for how you can reduce the data polling rate to just once per 10 seconds, instead of the default once per second.

```html
<div
  class="af-connect-module"
  data-on_response="onResponse"
  data-poll_rate="10000"
></div>
```

The table below shows all available configuration properties, default values and usage description.

| Configuration property      | Default value                     | Description                                                                          |
| --------------------------- | --------------------------------- | ------------------------------------------------------------------------------------ |
| data-label                  | AF Connect                        | Label displayed on the AF Connect Module interactive button.                         |
| data-poll_rate              | 1000                              | Data polling frequency, described in milliseconds.                                   |
| data-poll_retry             | 10                                | Data polling retry maximum count, e.g. if network connectivity has been lost.        |
| data-poll_timeout           | 300000                            | Data polling timeout, described in milliseconds.                                     |
| data-af_connect_url         | https://af-connect.local          | URL to AF-Connect service to open in new tab/window when the user clicks the button. |
| data-af_portability_url     | http://af-connect.local:8080      | URL to service where session token and user data can be obtained.                    |
| data-af_portability_api_key | dummykey                          | The API key used when querying for session token and polling for data.               |
| data-on_response            | undefined                         | Name of global callback function to call upon polling success/failure.               |
| data-on_warning             | `(code) => { console.warn(code)}` | Name of global callback function to call upon warning occurrence.                    |
| data-suppress_warnings      | false                             | Prevents warnings from being printed to console                                      |

## Build from source

```bash
git clone https://github.com/MagnumOpuses/af-connect-module.git
cd af-connect-module
npm install
npm run build
```

## Test

```bash
npm run test
```

## Error codes

| Code | Message                                                                       |
| ---- | ----------------------------------------------------------------------------- |
| E001 | Failed to fetch session token                                                 |
| E002 | Failed to fetch envelope                                                      |
| E003 | Polling interval exception                                                    |
| E004 | Polling interval timeout                                                      |
| E005 | Detected AF Connect Module incompatability with remote AF Portability service |
| E006 | Detected AF Connect Module incompatability with remote AF Connect service     |
| E007 | Compatability check with remote AF Connect service failed                     |
| E008 | Compatability check with remote AF Portability service failed                 |

## Built with

- NPM v6.4.1 (Node package manager)
- Browserify v16.5.0 (Javascript bundler for browser)

## Versions, current dev state and future

1.1.0-beta

## Contributing

We would love if you'd like to help us build and improve this product for the benefit of everyone. This project is intended to be a safe, welcoming space for collaboration, and contributors are expected to adhere to the [Contributor Covenant](http://contributor-covenant.org/) code of conduct.

Any contributions, feedback and suggestions are more than welcome.

Please read our guidelines for contribution [here](CONTRIBUTING_TEMPLATE.md).

## License

[Apache License 2.0](LICENSE.md)

## Acknowledgments

No acknowledgments yet.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.1.0-beta]

### Added

- Error codes: E005, E006, E007, E008
- Call to AF-Connect /checkCompatability upon start/initialization of this module.
- Call to AF-Portability /checkCompatability upon start/initialization of this module.
- New configuration property 'data-on_warning' for callback upon warning.
- New configuration property 'data-supress_warnings' to disable most warning messages.

## [1.0.2-beta]

### Fixed

- AF-Connect button now disabled while fetching session and polling for envelope. Button is re-activated upon successful envelope response, received error code or timeout.

### Added

- New configuration property "data-poll_retry" to set how many poll retry attempts to perform before exiting with error code, by default 10 retries.

### Changed

- Changed the default api-key to be "dummykey".

## [1.0.1]

### Added

- Codelist for error messages.
- Second argument `error` to `onResponse` callback invocations.


__test__/module.test.js
const mockAxios = require("axios");

test("getSession with mock api", () => {
  mockAxios.get.mockImplementationOnce(() =>
    Promise.resolve({ token: "aa31667d-afde-4941-8749-d71ff90bc247" })
  );

  const config = {
    afPortabilityUrl: "https://portabilitystage.arbetsformedlingen.se"
  };

  const connectModule = require("../lib/connect-module");
  return connectModule.getSession(config).then(token => {
    expect(token).toBeDefined();
    expect(typeof token).toBe("string");
    expect(token).toBe("aa31667d-afde-4941-8749-d71ff90bc247");
  });
});

test("getEnvelope with mock api", () => {
  mockAxios.get.mockImplementationOnce(() =>
    Promise.resolve(require("./cv.json"))
  );

  const onResponse = data => {
    console.log("Response: ", data);
  };

  const config = {
    pollRate: "1000", // 1 second
    timeout: "300000", // 5 minutes
    afConnectUrl: "https://demotest.arbetsformedlingen.se",
    afPortabilityUrl: "https://portabilitystage.arbetsformedlingen.se",
    onResponse: onResponse
  };

  const connectModule = require("../lib/connect-module");
  return connectModule
    .getEnvelope(config, "aa31667d-afde-4941-8749-d71ff90bc247")
    .then(cv => {
      expect(cv).toBeDefined();
      expect(typeof cv).toBe("object");
      expect(cv.documentId).toBeDefined();
      expect(cv.person).toBeDefined();
      expect(cv.profiles).toBeDefined();
    });
});


lib/connect-module.js
"use strict";

const axios = require("axios");

const ERROR_CODES = {
  E001: { code: "E001", message: "Failed to fetch session token" },
  E002: { code: "E002", message: "Failed to fetch envelope" },
  E003: { code: "E003", message: "Polling interval exception" },
  E004: { code: "E004", message: "Polling interval timeout" },
  E005: {
    code: "E005",
    message:
      "Detected AF Connect Module incompatability with remote AF Portability service"
  },
  E006: {
    code: "E006",
    message:
      "Detected AF Connect Module incompatability with remote AF Connect service"
  },
  E007: {
    code: "E007",
    message: "Compatability check with remote AF Connect service failed"
  },
  E008: {
    code: "E008",
    message: "Compatability check with remote AF Portability service failed"
  }
};

const getSession = config => {
  return axios
    .get(
      config.afPortabilityUrl + "/token?api-key=" + config.afPortabilityApiKey
    )
    .then(response => {
      return response.data.token;
    })
    .catch(err => {
      throw ERROR_CODES.E001;
    });
};

const getEnvelope = (config, session) => {
  return axios
    .get(config.afPortabilityUrl + "/cv?sessionToken=" + session)
    .then(response => {
      switch (response.data.status) {
        case 204:
          // Received no CV/Envelope ready, empty response
          return;
        default:
          // Envelope received!
          return response.data;
      }
    })
    .catch(err => {
      if (err.response) {
        switch (err.response.status) {
          case 500:
            return "Something went wrong... Code: " + err.response.status;
          default:
            return "Unknown response code: " + err.response.status;
        }
      }
      throw ERROR_CODES.E002;
    });
};

const fetchSequence = (config, button) => {
  button.style["background-color"] = "#AAAAAA";
  button.setAttribute("disabled", "");
  return getSession(config)
    .then(sessionToken => {
      return new Promise((resolve, reject) => {
        window
          .open(config.afConnectUrl + "?sessionToken=" + sessionToken, "_blank")
          .focus();

        let retry = 0;
        let retryMax = config.pollRetry;

        let timeoutId;
        let awaitingResponse = false;
        const timerId = setInterval(() => {
          // Only poll for envelope if there are no ongoing request.
          if (awaitingResponse) {
            return;
          }

          awaitingResponse = true;
          return getEnvelope(config, sessionToken)
            .then(cv => {
              awaitingResponse = false;
              retry = 0;

              if (cv !== undefined) {
                clearInterval(timerId);
                clearTimeout(timeoutId);
                if (config.onResponse) {
                  window[config.onResponse].call(undefined, cv);
                }
                resolve(cv);
              }
            })
            .catch(err => {
              awaitingResponse = false;
              if (retry < retryMax) {
                // Error occurred upon polling for envelope, retrying.
                retry++;
              } else {
                // Too many errors occurred upon polling for envelope, stop polling with error code.
                clearInterval(timerId);
                reject(err);
              }
            });
        }, config.pollRate);

        timeoutId = setTimeout(() => {
          clearInterval(timerId);
          reject(ERROR_CODES.E004);
        }, config.timeout);
      });
    })
    .then(() => {
      button.style["background-color"] = "#3040C4";
      button.removeAttribute("disabled");
    })
    .catch(err => {
      window[config.onResponse].call(undefined, undefined, err);
      button.style["background-color"] = "#3040C4";
      button.removeAttribute("disabled");
    });
};

const checkCompatability = config => {
  console.log("Checking compatability...");

  const checks = [];

  // Check compatability with portability
  checks.push(
    axios
      .get(config.afPortabilityUrl + "/actuator/health")
      .then(response => {
        return response.data;
      })
      .catch(err => {
        throw ERROR_CODES.E005;
      })
  );

  // Check compatability with af-connect
  checks.push(
    axios
      .get("http://af-connect.local:9801/health")
      .then(response => {
        return response.data;
      })
      .catch(err => {
        throw ERROR_CODES.E006;
      })
  );

  // Run all checks
  return Promise.all(checks).then(result => {
    console.log("All checks ran, result: ", result);
  });
};

const generateButton = config => {
  const button = document.createElement("button");
  button.appendChild(document.createTextNode(config.label));
  button.style["background-color"] = "#3040C4";
  button.style["color"] = "#eee";
  button.style["border"] = "0px";
  button.style["border-radius"] = "3px";
  button.style["padding"] = "6px 20px";
  button.style["font-weight"] = "600";
  button.addEventListener("click", evt => {
    fetchSequence(config, button);
  });
  return button;
};

module.exports = {
  CODE: ERROR_CODES,
  getSession,
  getEnvelope,
  fetchSequence,
  generateButton
};


index.js
"use strict";

const connectModule = require("./lib/connect-module");

const checkCompatability = (host, name, version, onIncompatible) => {
  const url = `${host}/checkCompatability?name=${name}&version=${version}`;
  return fetch(url, {
    method: "GET",
    mode: "cors"
  })
    .then(response => response.json())
    .then(body => {
      if (body.result === false) {
        if (onIncompatible !== undefined) {
          onIncompatible();
        }
      }
    });
};

const containers = document.getElementsByClassName("af-connect-module");
Array.prototype.forEach.call(containers, container => {
  container.style["display"] = "inline";

  let config = {
    name: "af-connect-module",
    version: "1.1.0-beta",
    label: container.getAttribute("data-label") || "AF Connect",
    pollRate: container.getAttribute("data-poll_rate") || "1000", // 1 second
    pollRetry: container.getAttribute("data-poll_retry") || "10",
    timeout: container.getAttribute("data-poll_timeout") || "300000", // 5 minutes
    afConnectUrl:
      container.getAttribute("data-af_connect_url") ||
      "https://af-connect.local",
    afPortabilityUrl:
      container.getAttribute("data-af_portability_url") ||
      "http://af-connect.local:8080",
    afPortabilityApiKey:
      container.getAttribute("data-af_portability_api_key") || "dummykey",
    onResponse: container.getAttribute("data-on_response") || undefined,
    onWarning:
      container.getAttribute("data-on_warning") ||
      (code => {
        console.warn(code);
      }),
    supressWarnings: container.getAttribute("data-suppress_warnings") || false
  };

  const button = connectModule.generateButton(config);
  container.appendChild(button);

  // Check compatability with remote af connect service
  checkCompatability(config.afConnectUrl, config.name, config.version, () => {
    if (!config.supressWarnings) {
      config.onWarning(connectModule.CODE.E006);
    }
  });

  // Check compatability with remote af portability service
  checkCompatability(
    config.afPortabilityUrl,
    config.name,
    config.version,
    () => {
      if (!config.supressWarnings) {
        config.onWarning(connectModule.CODE.E005);
      }
    }
  );
});
